cookbook/foreign_shell_scripts.md

```diff
@@ -0,0 +1,241 @@
+
+```nu
+let bash_greeting = '
+export GREETING="Hello";
+export FROM="from bash";
+'
+
+load-env (
+  $bash_greeting
+  | str trim
+  | lines
+  | parse 'export {name}="{value}";'
+  | transpose --header-row --as-record
+)
+
+print $"($env.GREETING) ($env.FROM)" # "Hello from bash"
+```
+
+This is perfectly fine for situations where you are sure of the exact format of
+the script and can predict parsing edge cases.
+
+This quickly gets tricky though, for example when the script is declaring a
+`PATH` variable that references its previous value
+(`export PATH="$PATH:/extra/path";`).
+
+There are ways to implement some form of expansion too, but at some point it
+might make more sense to leave the parsing to the shell it was meant for.
+
+## Capturing the environment from a foreign shell script
+
+A more complex approach is to run the script in the shell it is written for and
+do some hackery to capture the script's variables afterwards.
+
+Note: The following custom command assumes a Unix-like operating system, it may
+also be possible to create one for Windows that could capture variables from a
+PowerShell script.
+
+```nu
+# Returns a record of changed env variables after running a non-nushell script (passed via stdin), e.g. a bash script you want to "source"
+def capture-foreign-env [
+    --shell (-s): string = /bin/sh
+    # The shell to run the script in
+    # (has to support '-c' argument and POSIX 'env', 'echo', 'eval' commands)
+    --arguments (-a): list<string> = []
+    # Additional command line arguments to pass to the foreign shell
+] {
+    let script = $in;
+    let env_out = with-env { SCRIPT_TO_SOURCE: $script } {
+        ^$shell ...$arguments -c `
+        env
+        echo '<ENV_CAPTURE_EVAL_FENCE>'
+        eval "$SCRIPT_TO_SOURCE"
+        echo '<ENV_CAPTURE_EVAL_FENCE>'
+        env -u _ -u _AST_FEATURES -u SHLVL` # Filter out known changing variables
+    }
+    | split row '<ENV_CAPTURE_EVAL_FENCE>'
+    | {
+        before: ($in | first | str trim | lines)
+        after: ($in | last | str trim | lines)
+    }
+
+    # Unfortunate Assumption:
+    # No changed env var contains newlines (not cleanly parseable)
+    $env_out.after
+    | where { |line| $line not-in $env_out.before } # Only get changed lines
+    | parse "{key}={value}"
+    | transpose --header-row --as-record
+}
+```
+
+This runs a foreign shell script and captures the changed environment variables
+after running the script. This is done by parsing output of the `env` command
+available on unix-like systems. The shell to execute can be specified and
+configured using the `--shell` and `--arguments` parameters, the command has
+been tested using sh (bash), bash, zsh, fish, ksh, and dash.
+
+As an example, we can use this set up Homebrew paths using `brew shellenv`. Note
+that the path to the `brew` binary depends on the OS and installation.
+
+```
+load-env (^/opt/homebrew/bin/brew shellenv | capture-foreign-env)
```

Comment:
This is perhaps the most important part of the page: A quick example of how to use the capture command. Ideally, it should be something more widely understandable. People not using Homebrew don't know what `brew shellenv` is. I think even `load-env ('script.sh' | capture-foreign-env)` would do. It is not immediately clear the `$in` should be a script filename, some people might try to do `open script.sh | capture-foreign-env` or similar to pipe in the script contents.

Reply:
(By "people" I mean "me" as these things came to mind when reading it.)

Reply:
Yeah, that makes a lot of sense! I'll create a simpler, highlighted example tomorrow and fix a few leftover typos.

Reply:
@kubouch Actually, after rereading your comment: I didn't plan for you to be able to specify a script name, and I don't think it works, unless I missed something about `eval`.

So, `open` or `open --raw` would be the way to go, when having an existing script file. I liked having only one input method and thought it's important to allow the piping-in of the contents because tools generating env-var-scripts often pipe the script to stdout. Like brew in this case.

But I'm up for suggestions!
For now I'd add the example with `open`

Reply:
OK, then I misunderstood. It's not clear whether `$in` should be a script name or a source string, I didn't read the script source that thoroughly. Then the example could be `load-env (open script.nu | capture foreign-env)` or similar.

Side note: You can check [bass](https://github.com/edc/bass) if you didn't know it. A similar idea but for running bash scripts in fish, written in Python.

Reply:
Oh I see, no worries!
I updated the PR and tried to be a little more clear about this.